chore: Cruise control refactor

samuel-hawker opened the pull request:
### Type of change

- Refactoring

### Description

This PR is a minor cleanup and consolidation of names, and slight tweaks to test comments to more accurately describe and specify the objects and states involved.

IN order to clean up several JSON related tests I have also created a JSONObjectMatchers class that contains custom hamcrest matchers that reduces the overall test code.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -413,29 +434,29 @@ private KafkaRebalanceStatus buildRebalanceStatus(String sessionID, KafkaRebalan
-        if (rebalanceAnnotation == KafkaRebalanceAnnotation.refresh) {
-            // the user fixed the error on the resource and want to "refresh", actually
-            // requesting a new rebalance proposal
+        if (KafkaRebalanceAnnotation.refresh.equals(rebalanceAnnotation)) {
```
the `equals` implementation for enum is just `this == other`, tbh I think that using `equals` for enums is less readable. Furthermore, I found articles talking about preferring `==` instead of `equals` just for enums.

ppatierno replied:
@samuel-hawker do you think to address this comment?

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -455,66 +477,65 @@ private KafkaRebalanceStatus buildRebalanceStatus(String sessionID, KafkaRebalan
-                                                }
-                                            });
+                                            currentKafkaRebalance.getStatus().getSessionId())
+                                        .onFailure(e -> {
```
can we put the `onFailure` after the `onSuccess` just for having branch code that should be called most of the times as first during the code reading?

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -564,25 +585,25 @@ private KafkaRebalanceStatus buildRebalanceStatus(String sessionID, KafkaRebalan
         Promise<KafkaRebalanceStatus> p = Promise.promise();
-        if (rebalanceAnnotation == KafkaRebalanceAnnotation.none) {
-            log.info("{}: Arming Cruise Control rebalance user task status timer", reconciliation);
+        if (KafkaRebalanceAnnotation.none.equals(rebalanceAnnotation)) {
```
ditto as above

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -591,31 +612,32 @@ private KafkaRebalanceStatus buildRebalanceStatus(String sessionID, KafkaRebalan
-                            if (state(freshKafkaRebalance) == KafkaRebalanceState.Rebalancing) {
-                                if (rebalanceAnnotation(freshKafkaRebalance) == KafkaRebalanceAnnotation.stop) {
+                kafkaRebalanceOperator.getAsync(kafkaRebalance.getMetadata().getNamespace(), kafkaRebalance.getMetadata().getName())
+                    .onFailure(e -> {
```
ditto as above

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -794,36 +814,36 @@ private KafkaRebalanceStatus buildRebalanceStatus(String sessionID, KafkaRebalan
-                        if (response.thereIsNotEnoughDataForProposal()) {
-                            // If there are not enough data for a rebalance, it's an actual error at Cruise Control level
-                            // and we need to re-request the proposal at a later stage so we move to the PendingProposal State.
+                        if (!response.isSufficientDataForProposal()) {
```
I think that @tomncooper was using the logic with `thereIsNotEnough....` because the Cruise Control has something similar about the window in which you need samples for data. I would leave that logic tbh to have the same kind of mapping with Cruise Control logic. Starting to use the opposite logic in the code and then negate it ... it's quite confusing. Just my 2 cents.

tomncooper replied:
The "NotEnoughData" flag is based on the "NotEnoughValidWindows" error on the CC side. It is an exceptional state (hopefully) so we went with a flag that has to be actively set to true when the error occurs. I am not clear what the advantage is to swapping this around?

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlRebalanceResponse.java:
```diff
@@ -8,28 +8,30 @@
 
-    private boolean notEnoughDataForProposal;
-    private boolean proposalIsStillCalculating;
+    private boolean isSufficientDataForProposal;
```
as already mentioned above I think we should leave the logic `notEnoughData...` as it is in Cruise Control.

ppatierno commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperatorTest.java:
```diff
@@ -250,60 +249,60 @@ public void testNewToPendingProposalToStoppedRebalance(Vertx vertx, VertxTestCon
                 public void run() {
                     if (++count == 4) {
-                        // after a while, apply the "stop" annotation to the resource in the PendingProposal state
+                        // After a while, apply the "stop" annotation to the resource in the PendingProposal state
```
it's a personal choice ... I always start comments with low letter :-) otherwise you should fix in a TONS of places :-)

ppatierno commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/MockCruiseControl.java:
```diff
@@ -70,24 +70,28 @@ public static ClientAndServer getCCServer(int port) throws IOException, URISynta
-        ClientAndServer ccServer = new ClientAndServer(port);
-        return ccServer;
+        ClientAndServer server = new ClientAndServer(port);
+        return server;
```
at this point even just `return new ClientAndServer(port);`

tomncooper commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -455,66 +477,65 @@ private KafkaRebalanceStatus buildRebalanceStatus(String sessionID, KafkaRebalan
-                            // and the previous execution set the status and completed the future
-                            if (state(freshKafkaRebalance) == KafkaRebalanceState.PendingProposal) {
-                                if (rebalanceAnnotation(freshKafkaRebalance) == KafkaRebalanceAnnotation.stop) {
+        if (KafkaRebalanceAnnotation.none.equals(rebalanceAnnotation)) {
```
I agree with @ppatierno, that the "==" version reads cleaner that this ".equals" version

tomncooper commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -794,36 +814,36 @@ private KafkaRebalanceStatus buildRebalanceStatus(String sessionID, KafkaRebalan
                             return buildRebalanceStatus(null, KafkaRebalanceState.PendingProposal);
-                        } else if (response.proposalIsStillCalculating()) {
-                            // If rebalance proposal is still being processed we need to re-request the proposal at a later stage
+                        } else if (response.isProposalInProgress()) {
```
Most of the time a user will not be asking for some exotic rebalance with strange options and will just get the cached proposal straight back or close to it. The "proposalStillCalaculating" state is assumed to be exceptional so that is why it a flag that has to be explicitly set to true. I am not sure what the advantage is in switching this around.

tomncooper commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/MockCruiseControl.java:
```diff
@@ -70,24 +70,28 @@ public static ClientAndServer getCCServer(int port) throws IOException, URISynta
     private static JsonBody getJsonFromResource(String resource) throws URISyntaxException, IOException {
 
-        URI jsonURI = Objects.requireNonNull(MockCruiseControl.class.getClassLoader().getResource(CC_JSON_ROOT + resource)).toURI();
+        URI jsonURI = Objects.requireNonNull(MockCruiseControl.class.getClassLoader().getResource(CC_JSON_ROOT + resource))
```
Out of interest what hard-wrap width are you working to, seems a lot of edits in this PR are hard wraps. I actually think it would be nice to define a project wide one as a style guide.

samuel-hawker replied:
I tend to wrap in logical sections, readbaility is unfortunately a little subjective but the reason I split this was to make clear that we are modifying to the objects, then converting them to a URI
i.e. one clear operation per a line, for things like filters this rule can get a bit more subjective and arguable, but just because a final method is a simple conversioon and only a few letters in length for the method I don't neccessarily think it belongs on the same line as the rest, if that makes sense.

ultimately though this is styling and consistency only has so much value.
Would be open to discussing a hardwrap limit though!

samuel-hawker commented:
I think your comments are very reasonable, generally I don't like the `isNot...` as a boolean as it causes confusion in my experience, and that the 'good condition' should generally be true.
Maybe this is a convention that is not necessarily shared by everyone, happy to change it back if you think there is more confusion in mappoing the CC notEnough to the isSufficient

michalxo commented on operator-common/src/main/java/io/strimzi/operator/common/Annotations.java:
```diff
@@ -20,8 +20,8 @@
-    // this annotation with related possible values (approve, stop, refresh) is set by the user for interacting
-    // with the rebalance operator in order to start, stop or refresh rebalacing proposals and operations
+    // This annotation with related possible values (approve, stop, refresh) is set by the user for interacting
+    // with the rebalance operator in order to start, stop or refresh rebalancing proposals and operations
```
```suggestion
    // with the rebalance operator in order to start, stop or refresh rebalancing proposals and operations.
```

kyguy replied:
stop -> stop,

(I think there there should be a comma there, if not, please ignore this comment!)

kyguy commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -720,62 +736,66 @@ private KafkaRebalanceStatus buildRebalanceStatus(String sessionID, KafkaRebalan
+                                        + "' identified by label '" + Labels.STRIMZI_CLUSTER_LABEL
+                                        + "' does not exist in namespace " + clusterNamespace + ".")).mapEmpty();
+                    } else if (kafka.getSpec().getCruiseControl() == null) {
+                        log.warn("{}: Kafka resouce lacks 'cruiseControl' declaration : No deployed Cruise Control for doing a rebalance.", reconciliation);
```
resouce -> resource

kyguy commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -720,62 +736,66 @@ private KafkaRebalanceStatus buildRebalanceStatus(String sessionID, KafkaRebalan
+                    } else if (kafka.getSpec().getCruiseControl() == null) {
+                        log.warn("{}: Kafka resouce lacks 'cruiseControl' declaration : No deployed Cruise Control for doing a rebalance.", reconciliation);
+                        return updateStatus(kafkaRebalance, new KafkaRebalanceStatus(),
+                                new InvalidResourceException("Kafka resouce lacks 'cruiseControl' declaration "
```
resouce -> resource

kyguy commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -720,62 +736,66 @@ private KafkaRebalanceStatus buildRebalanceStatus(String sessionID, KafkaRebalan
+                        .compose(currentKafkaRebalance -> {
+                            KafkaRebalanceStatus kafkaRebalanceStatus = currentKafkaRebalance.getStatus();
+                            KafkaRebalanceState currentState;
+                            // cluster rebalance is new or it is in one of others states
```
of others states -> of the other states

kyguy commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -720,62 +736,66 @@ private KafkaRebalanceStatus buildRebalanceStatus(String sessionID, KafkaRebalan
-                                                + ": No deployed Cruise Control for doing a rebalance.")).mapEmpty();
+                                String rebalanceStateType = rebalanceStateConditionType(kafkaRebalanceStatus);
+                                if (rebalanceStateType == null) {
+                                    throw new RuntimeException("Unable to find KafkaRebalace State in current KafkaRebalance status");
```
KafkaRebalace -> KafkaRebalance

kyguy commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java:
```diff
@@ -58,19 +57,20 @@
- *     This operator takes care of the {@code KafkaRebalance} custom resources that a user can create in order
- *     to interact with Cruise Control REST API and execute a cluster rebalancing.
+ *     This operator takes care of the {@code KafkaRebalance} custom resource type that a user can create in order
+ *     to interact with the Cruise Control REST API and execute a cluster rebalancing.
```
cluster rebalancing -> cluster rebalance

samuel-hawker commented:
I believe I have addressed all comments, and added extra documentation domments to the CruiseControlResponse class to explain why the booleans are named as they are.

samuel-hawker commented:
I will add that the `isProposalCalculating` is actually set to true if the `progress` key is set in the response, not explicitly checking the value of the progress key. The tests suggest that this is the desired behaviour, but I do wonder if the progress key could ever be present and say they are all complete?
If so a check of the inner object might be useful.But for now if no troubles have ever been encountered leaving it as is would probably be the best.

scholzj commented:
@ppatierno Could you have a look if this looks ok now for you and your comments were addressed?

ppatierno commented:
I had another pass I haven't seen the enums comparison to be fixed. I was wondering if it's intentional or not @samuel-hawker

samuel-hawker commented:
/azp run acceptance

im-konge commented:
@strimzi-ci run tests test_only

im-konge commented:
It seems that Jenkins will not work today too, the only failing test for acceptance profile on azure pipelines is not related to the PR. @samuel-hawker I think that's ready for merge if someone else doesn't want to add some review.

scholzj commented:
/azp run acceptance

scholzj commented:
The `testRecoveryFromKafkaStatefulSetDeletion` failure is not related to this. There is another PR to fix that.

@kyguy @michalxo You had comments to this PR in the past. Is this ok to merge now?

michalxo reviewed (approved):
No problems from my side.
